thriftbp/circuit_breaker.go

```diff
@@ -0,0 +1,102 @@
+package thriftbp
+
+import (
+	"context"
+	"time"
+
+	"github.com/apache/thrift/lib/go/thrift"
+	"github.com/reddit/baseplate.go/log"
+	"github.com/reddit/baseplate.go/metricsbp"
+	"github.com/sony/gobreaker"
+)
+
+// FailureRatioBreaker is a circuit breaker based on gobreaker that uses a low-water-mark and
+// % failure threshold to trip.
+type FailureRatioBreaker struct {
+	*gobreaker.CircuitBreaker
+
+	metricPrefix      string
+	minRequestsToTrip int
+	failureThreshold  float64
+}
+
+// CircuitBreakerConfig represents the configuration for a FailureRatioBreaker
+type CircuitBreakerConfig struct {
+	// Minimum requests that need to be sent during a time period before the breaker is eligible to transition from closed to open.
+	MinRequestsToTrip int
+	// Percentage of requests that need to fail during a time period for the breaker to transition from closed to open.
+	// Represented as a float in {0,1}, where .05 means >5% failures will trip the breaker.
+	FailureThreshold float64
+	// Gobreaker settings used to configure the underlying circuit breaker.
+	Settings gobreaker.Settings
+}
+
+// NewFailureRatioBreaker creates a new FailureRatioBreaker with the provided configuration.
+func NewFailureRatioBreaker(config CircuitBreakerConfig) FailureRatioBreaker {
+
+	failureBreaker := FailureRatioBreaker{
+		metricPrefix:      config.Settings.Name,
+		minRequestsToTrip: config.MinRequestsToTrip,
+		failureThreshold:  config.FailureThreshold,
+	}
+	config.Settings.OnStateChange = stateChanged
+	config.Settings.ReadyToTrip = failureBreaker.ShouldTripCircuitBreaker
+	failureBreaker.CircuitBreaker = gobreaker.NewCircuitBreaker(config.Settings)
+	go failureBreaker.runStatsProducer()
+	return failureBreaker
+}
+
+func (cb FailureRatioBreaker) runStatsProducer() {
+	circuitBreakerGauge := metricsbp.M.RuntimeGauge(cb.metricPrefix + ".circuit.breaker.closed")
```

Comment:
nit: slightly prefer `".circuit-breaker-closed"` over `".circuit.breaker.closed"`. We are not gaining anything from adding 2 extra layers of the metrics path.